File: 백준/Silver/9012. 괄호/괄호.py
import sys
T=int(input())
for _ in range(T):
    s = sys.stdin.readline().strip()
    stack=[]
    tf = True
    if len(s)%2 == 1:
        print('NO')
        continue
    else:
        for i in s:
            if i=='(':
                stack.append('(')
            else :
                if len(stack)>0:
                    stack.pop()
                else:
                    tf = False
            
    if len(stack)!=0 or tf == False:
        print('NO')
    else :
        print('YES')
      

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            stack.clear();
            boolean tf = true;
            String str = br.readLine();
            for (int j = 0; j < str.length(); j++) {
                if(str.charAt(j)=='('){
                    stack.add(str.charAt(j));
                }
                if(str.charAt(j)==')'){
                    if(!stack.isEmpty()){
                        stack.pop();
                    }else{
                        tf = false;
                        break;
                    }
                }
            }

            if(tf){
                if(!stack.isEmpty()){
                    sb.append("NO\n");
                }else{
                    sb.append("YES\n");
                }
            }else{
                sb.append("NO\n");
            }
        }
        System.out.println(sb);
    }
}